source/common/event/file_event_impl.h

```diff
@@ -41,5 +45,119 @@ class FileEventImpl : public FileEvent, ImplBase {
   // polling and activating new fd events.
   const bool activate_fd_events_next_event_loop_;
 };
+
+// Forward declare for friend class.
+class UserSpaceFileEventFactory;
+
+// The interface of populating event watcher and obtaining the active events. The events includes
+// Read, Write and Closed.
+class EventListener {
+public:
+  virtual ~EventListener() = default;
+
+  // Provide the activated events.
+  virtual uint32_t triggeredEvents() PURE;
+  virtual uint32_t getAndClearEphemeralEvents() PURE;
+
+  // Callbacks of the event operation.
+  virtual void onEventEnabled(uint32_t enabled_events) PURE;
+  virtual void onEventActivated(uint32_t activated_events) PURE;
+};
+
+// Return the enabled events except EV_CLOSED. This implementation is generally good since only
+// epoll supports EV_CLOSED but the entire envoy code base supports another poller. The event owner
+// must assume EV_CLOSED is never activated. Also event owner must tolerate that OS could notify
```

Comment:
why not closed?
should we ASSERT it's not enabled?

Reply:
> why not closed? should we ASSERT it's not enabled?

I need a complicated EventListener implementation to supporting CLOSED event.
EventListener need to consume byte transmition status while read/write doesn't.

EV_CLOSED is not a MUST-have but a nice-to-have is that IIRC libevent only deliver EV_CLOSED if the poller is epoll.

I add the TODO for now